logstash-core/spec/logstash/pipelines_registry_spec.rb

```diff
@@ -101,6 +101,39 @@
         end
       end
     end
+
+    context "when pipeline is initializing" do
+      let (:wait_start_create_block) { Queue.new }
+      let (:wait_before_exiting_create_block) { Queue.new }
+      let (:slow_initializing_pipeline) { double("slow_initializing_pipeline") }
+      let (:pipeline2) { double("pipeline2") }
+
+      it "should create a loading state before calling the create block" do
+        t = Thread.new do
+          subject.create_pipeline(pipeline_id, slow_initializing_pipeline) do
+            # signal that we entered the create block
+            wait_start_create_block << "ping"
+
+            # stall here until wait_before_exiting_create_block receives a message
+            wait_before_exiting_create_block.pop
+
+            true
+          end
+        end
+
+        # stall here until subject.create_pipeline has been called and it entered the create block
+        wait_start_create_block.pop
+
+        expect(slow_initializing_pipeline).not_to receive(:finished_execution?)
```

Comment:
This is rather complicated sequence to test a rather complicated scenario.
what do you think we can do to help readers understand this test quicker?

Reply:
added more comments